### ISSUE TITLE
AdventureZ prevents player from swimming 

### ISSUE THREAD

**Issue by YShadowMY** [Sinytra/Connector#1343]
### Mod Name

AdventureZ

### Mod Homepage

https://www.curseforge.com/minecraft/mc-mods/adventurez

### Minecraft version

1.21.1

### Describe the bug

while the player is in the water, he cannot move in any direction

### Steps to reproduce

1. download Connector, AdventureZ + dependencies
2. create a new world 
3. try swimming in the water

### Logs

https://gist.github.com/YShadowMY/08c11f64e718f8210532fcd341af4a3e

**Comment by Tursiops-G** [comment]
Same exact problem with UselessReptile... Player Can't move in water. (everything else works, AFAIK).

**Comment by Cixon** [comment]
confirmed bug.

**Comment by starrysais** [comment]
this also happens with betterend and its dependencies.

**Comment by Cixon** [comment]
> this also happens with betterend and its dependencies.

i would open a separate ticket for this !

**Comment by JalinRio42** [comment]
> This also happens with betterend and its dependencies.

Had this same issue. It was just better end for me. 
It also affects all mobs. which was super odd watching pigs drown and fish just float in one spot.

**Comment by Cixon** [comment]
bump

**Comment by coco875** [comment]
> bump

bump are not useful, the dev work on something else for now so just wait

**Comment by SmexyJesse** [comment]
I got a work around.
-Disable the problem mods and create a new world
-Check swimming works
-Save and exit
-Enable the problem mods
-Load back into new world you created
-Should be able to swim with the problem mods

**Comment by Cixon** [comment]
yeah but not an ideal work around. for modpack creators, downloaders expect things to just work.

> I got a work around.
> -Disable the problem mods and create a new world
> -Check swimming works
> -Save and exit
> -Enable the problem mods
> -Load back into new world you created
> -Should be able to swim with the problem mods

**Comment by SmexyJesse** [comment]
> yeah but not an ideal work around. for modpack creators, downloaders expect things to just work.
> 
> > I got a work around.
> > -Disable the problem mods and create a new world
> > -Check swimming works
> > -Save and exit
> > -Enable the problem mods
> > -Load back into new world you created
> > -Should be able to swim with the problem mods

yeah but this information can help people who wanna play their own mod packs and help devs locate the cause of the bug.

**Comment by Atrimilan** [comment]
I think I've found the variable that needs to be fixed, it's in the world file `level.dat`, which you can edit with [NBTExplorer](https://github.com/jaquadro/NBTExplorer).
> You don't need to close Minecraft for this process, just quit your world and do the following.

* Open NBTExplorer
* Select the world in which you can't swim, go to `level.dat` > `Data` > `Player` > `attributes`
* There should be a few entries, and one of them contains `id: neoforge:swim_speed`
* So, in this entry, double-click on `base: 0` and replace it with `base: 1`
  ![Image](https://github.com/user-attachments/assets/809fa1f5-7781-45a2-b371-d82acebae838)
* Save
* Check if this has solved the swimming problem

**Comment by Tursiops-G** [comment]
Tried the NBTExplorer trick, No joy.
The id: neoforge:swim_speed attribute tag entry didn't exist.
Tried creating/saving it, but it gets overwritten/deleted when subsequently reloading the modified world save. 

Note: this is using Useless Reptile 0.8.3-1.21.1 (Fabric mod).

**Comment by Atrimilan** [comment]
Yes, I added Useless Reptile and other mods, and my workaround no longer works...

I also have another bug, probably related somehow, when I fly in creative mode and I change handheld item, the gravity reactivates and I can't fly anymore (I have to switch to spectator mode to reactivate it).
> I don't have AdventureZ, but I have BetterEnd, BetterNether, Useless Reptile, RPG Series mods, and other small mods.

**Comment by unilock** [comment]
> I think I've found the variable that needs to be fixed, it's in the world file `level.dat`, which you can edit with [NBTExplorer](https://github.com/jaquadro/NBTExplorer).

The in-game command equivalent would be `/attribute <username> neoforge:swim_speed base set 1.0`, but that didn't work for me either.

**Comment by cocoa005** [comment]
> > I think I've found the variable that needs to be fixed, it's in the world file `level.dat`, which you can edit with [NBTExplorer](https://github.com/jaquadro/NBTExplorer).
> 
> The in-game command equivalent would be `/attribute <username> neoforge:swim_speed base set 1.0`, but that didn't work for me either.

This method (the command) is actually successful in resolving the issue for the mod I'm trying to use at least. Thank you for sharing it.
Edit: Although it should be noted that this only resolves the issue for the player, not other non-aquatic entities.

**Comment by unilock** [comment]
I've created a workaround in the form of a mod: https://github.com/unilock/sinytra1343

All it does is make `LivingEntity#getAttributeValue`, when its parameter is `NeoForgeMod.SWIM_SPEED`, always return `1.0`.  
Note that this will obviously break other mods that may change that attribute's value, but at least it restores vanilla behavior.

**Comment by Cixon** [comment]
> I've created a workaround in the form of a mod: https://github.com/unilock/sinytra1343
> 
> All it does is make `LivingEntity#getAttributeValue`, when its parameter is `NeoForgeMod.SWIM_SPEED`, always return `1.0`. Note that this will obviously break other mods that may change that attribute's value, but at least it restores vanilla behavior.

goated. i will need to try it in my modpack to see if there’s any strange behaviors

**Comment by Atrimilan** [comment]
> I've created a workaround in the form of a mod: https://github.com/unilock/sinytra1343
> 
> All it does is make `LivingEntity#getAttributeValue`, when its parameter is `NeoForgeMod.SWIM_SPEED`, always return `1.0`. Note that this will obviously break other mods that may change that attribute's value, but at least it restores vanilla behavior.

Swimming seems to work fine with your workaround in my local world, thanks 😄.

**Comment by Atrimilan** [comment]
> I also have another bug, probably related somehow, when I fly in creative mode and I change handheld item, the gravity reactivates and I can't fly anymore (I have to switch to spectator mode to reactivate it).

In case anyone else is having the same flying issue in creative mod, simply update @unilock's workaround like so :
```java
return attribute.is(NeoForgeMod.SWIM_SPEED::is) || attribute.is(NeoForgeMod.CREATIVE_FLIGHT::is) ? 1.0D : original;
```
You'll still lose your flight when changing handheld items, but you'll be able to fly again without changing game mode.

**Comment by Cixon** [comment]
hey @unilock is it possible for you to make a mod/patch for issue #1435 ? I reached out to mod creator of Legacy 4J and they mentioned that connector is not allowing the nametag rendering.

## Please and thank you for everything!

**Comment by Atrimilan** [comment]
> hey [@unilock](https://github.com/unilock) is it possible for you to make a mod/patch for issue [#1435](https://github.com/Sinytra/Connector/issues/1435) ? I reached out to mod creator of Legacy 4J and they mentioned that connector is not allowing the nametag rendering.

Maybe you can try updating the code by changing the `neoforge:nametag_distance` ?

Because without any mod, I get:
![Image](https://github.com/user-attachments/assets/36258415-d88d-43c1-b6d5-ce5c2fb100bc)
And with betterend, etc. I get `0.0`.

I've seen this property here too : https://nekoyue.github.io/ForgeJavaDocs-NG/javadoc/1.20.6-neoforge/net/neoforged/neoforge/common/NeoForgeMod.html#NAMETAG_DISTANCE

Maybe you can simply edit the unilock workaround to something like the following:

```java
        if (attribute.is(NeoForgeMod.SWIM_SPEED::is) || attribute.is(NeoForgeMod.CREATIVE_FLIGHT::is)) {
            return 1.0D;
        } else if (attribute.is(NeoForgeMod.NAMETAG_DISTANCE::is)) {
            return 64.0D;
        } else {
            return original;
        }
```

I don't know if it's actually the affected attribute, I haven't tried it yet.

**Comment by unilock** [comment]
I've updated my mod to only activate when the detected attribute value is `0.0` (so it may be slightly more compatible with other mods), and to account for `NeoForgeMod.NAMETAG_DISTANCE` (returns `64.0`) and `NeoForgeMod.CREATIVE_FLIGHT` (returns `1.0` if the `LivingEntity` is a `Player` and the player is in creative mode).

You can download it here: https://github.com/unilock/sinytra1343/releases/tag/1.1.0

**Comment by Cixon** [comment]
goated. for now, i'm trying to help the community get BetterEnd stuff to work, i have a fix... but i need help.
https://github.com/Sinytra/Connector/issues/1331#issuecomment-2467050681

**Comment by Tursiops-G** [comment]
> I've updated my mod to only activate when the detected attribute value is `0.0` (so it may be slightly more compatible with other mods), and to account for `NeoForgeMod.NAMETAG_DISTANCE` (returns `64.0`) and `NeoForgeMod.CREATIVE_FLIGHT` (returns `1.0` if the `LivingEntity` is a `Player` and the player is in creative mode).
> 

Swimming bug fix Confirmed working w/ Useless reptile (Using Latest Neo & FFAPI as well)... Thank you very much for your work on this. :)

**Comment by Cixon** [comment]
I tried to get Adventure Z to work with newer version of NeoForge for instance 21.1.72 and it won't work. I know for a fact the older NeoForge version it works but something changed. Is anyone else experiencing this?

**Comment by Cixon** [comment]
> I tried to get Adventure Z to work with newer version of NeoForge for instance 21.1.72 and it won't work. I know for a fact the older NeoForge version it works but something changed. Is anyone else experiencing this?

https://github.com/Sinytra/Connector/issues/1527

**Comment by horse4lunch** [comment]
I was having this same issue when I updated my server. I was able to fix by running this command "/attribute @s neoforge:swim_speed base set 1"

**Comment by Maeiro** [comment]
> I've updated my mod to only activate when the detected attribute value is `0.0` (so it may be slightly more compatible with other mods), and to account for `NeoForgeMod.NAMETAG_DISTANCE` (returns `64.0`) and `NeoForgeMod.CREATIVE_FLIGHT` (returns `1.0` if the `LivingEntity` is a `Player` and the player is in creative mode).
> 
> You can download it here: https://github.com/unilock/sinytra1343/releases/tag/1.1.0

Hi! Just dropping by to share that I encountered the same issue with swimming after installing additional mods besides BetterEnd.

In my case, SWIM_SPEED wasn't 0.0D, but a very low value like 0.07D, which was still breaking swimming movement.

I made a small adjustment to the fix by also handling low values (≤ 0.1D) instead of only checking for exactly 0.0D.

I've tested it and, so far, it seems to have solved the problem for me.

I've also opened a pull request with the improvement in case it helps others. Thanks again for the initial patch!

![Image](https://github.com/user-attachments/assets/e91a9a8c-9e88-4bd2-ad5f-e2006ac63cc2)

**Comment by Maeiro** [comment]
To those that are not sure whether they're affected by this issue:

You can quickly check your SWIM_SPEED attribute directly in-game while you're in water by running the following command:

`/attribute @p neoforge:swim_speed get`

If it returns a very low value like 0.07D, that's likely the reason you're unable to swim properly.

Ideally, it should be closer to 1.0D for normal swimming behavior.

If you see a low value, the fix I suggested (handling values ≤ 0.1D instead of only 0.0D) should solve the problem.

**Comment by elijah-cul** [comment]
same issue with betterend

what should i do if i have 0.0D?

**Comment by unilock** [comment]
> In my case, SWIM_SPEED wasn't 0.0D, but a very low value like 0.07D, which was still breaking swimming movement.

This issue has been accounted for with [v1.2.0](https://github.com/unilock/sinytra1343/releases/tag/1.2.0) of my fix mod :)

**Comment by unilock** [comment]
I injected the following at the HEAD of `LazyEntityAttributes#lambda$updateAttributeSuppliers$3` using Mixin to do some debugging:

```
if (value.hasAttribute(placeholder)) {
	System.out.println("AttributeSupplier has placeholder for "+originalSupplier.getRegisteredName()+". Replacing with original!");
} else {
	System.out.println("AttributeSupplier does not have placeholder for "+originalSupplier.getRegisteredName()+".");
}
```

The following was printed to the log while starting the game:

<details>

```
AttributeSupplier has placeholder for artifacts:generic.mount_speed. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.sprinting_step_height. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.sprinting_speed. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.flatulence. Replacing with original!
AttributeSupplier does not have placeholder for artifacts:player.entity_experience.
AttributeSupplier does not have placeholder for artifacts:player.villager_reputation.
AttributeSupplier has placeholder for artifacts:generic.invincibility_ticks. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.slip_resistance. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.attack_burning_duration. Replacing with original!
AttributeSupplier has placeholder for artifacts:generic.movement_speed_on_snow. Replacing with original!
AttributeSupplier does not have placeholder for neoforge:swim_speed.
AttributeSupplier has placeholder for artifacts:generic.drinking_speed. Replacing with original!
AttributeSupplier does not have placeholder for neoforge:nametag_distance.
AttributeSupplier has placeholder for artifacts:generic.eating_speed. Replacing with original!
AttributeSupplier does not have placeholder for neoforge:creative_flight.
AttributeSupplier has placeholder for artifacts:player.entity_experience. Replacing with original!
AttributeSupplier has placeholder for artifacts:player.villager_reputation. Replacing with original!
AttributeSupplier does not have placeholder for artifacts:generic.mount_speed.
AttributeSupplier does not have placeholder for artifacts:generic.sprinting_step_height.
AttributeSupplier does not have placeholder for artifacts:generic.sprinting_speed.
AttributeSupplier does not have placeholder for artifacts:generic.flatulence.
AttributeSupplier does not have placeholder for artifacts:generic.invincibility_ticks.
AttributeSupplier does not have placeholder for artifacts:generic.slip_resistance.
AttributeSupplier does not have placeholder for artifacts:generic.attack_burning_duration.
AttributeSupplier does not have placeholder for artifacts:generic.movement_speed_on_snow.
AttributeSupplier does not have placeholder for artifacts:generic.drinking_speed.
AttributeSupplier does not have placeholder for artifacts:generic.eating_speed.
```
</details>

(many times, presumably once for each EntityType - duplicate lines were permuted and removed)

So it looks like the placeholders for NeoForge's attributes are never being added to the AttributeSuppliers!?

EDIT: Further discussion is taking place in Sinytra's Discord guild.

**Comment by unilock** [comment]
I've updated my fix mod to use a much better approach to fixing the problem. It should now work fully as intended, including with curios and other items with attribute modifiers.

https://github.com/unilock/sinytra1343/releases/tag/2.0.0

**Comment by SwiftYeeter10** [comment]
**For those finding this thread from search**
Until this gets fixed theres a patch mod available **Here**: [https://modrinth.com/mod/sinytra1343](url).
